lib/rex/proto/dns/static_hostnames.rb

```diff
@@ -0,0 +1,151 @@
+# -*- coding: binary -*-
+
+require 'rex/socket'
+require 'forwardable'
+
+module Rex
+module Proto
+module DNS
+  ##
+  # This class manages statically defined hostnames for DNS resolution where each name is a mapping to an IPv4 and or
+  # an IPv6 address. A single hostname can only map to one address of each family.
+  ##
+  class StaticHostnames
+    extend Forwardable
+
+    def_delegators :@hostnames, :each, :each_with_index, :length, :empty?
+
+    # @param [Hash<String, IPAddr>] hostnames The hostnames to IP address mappings to initialize with.
+    def initialize(hostnames: nil)
+      @hostnames = {}
+      if hostnames
+        hostnames.each do |hostname, ip_address|
+          add(hostname, ip_address)
+        end
+      end
+    end
+
+    # Locate and parse a hosts file on the system. Only the first hostname to IP address definition is used which
+    # replicates the behavior of /etc/hosts on Linux. Loaded definitions are merged with existing definitions.
+    def parse_hosts_file
+      path = %w[
+        %WINDIR%\system32\drivers\etc\hosts
+        /etc/hosts
+        /data/data/com.termux/files/usr/etc/hosts
+      ].find do |path|
+        path = File.expand_path(path)
+        File.file?(path) && File.readable?(path)
+      end
+      return unless path
+
+      path = File.expand_path(path)
+      ::IO.foreach(path) do |line|
+        words = line.split
+        next unless words.length > 1 && Rex::Socket.is_ip_addr?(words.first)
+
+        ip_address = IPAddr.new(words.shift)
+        words.each do |hostname|
+          add(hostname, ip_address)
+        end
+      end
+    end
+
+    # Get an IP address of the specified type for the hostname. Only the first address is returned in cases where
+    # multiple addresses are defined.
```

Comment:
Not a blocker; It'd be good to document any guarantees here, i.e. if it's an ordered result or non-deterministic